src/tools/compiletest/src/main.rs

```diff
@@ -524,3 +526,24 @@ fn extract_lldb_version(full_version_line: Option<String>) -> Option<String> {
     }
     None
 }
+
+fn detect_node() -> Option<String> {
+    // Look for `nodejs` or `node` on PATH in that order, returning the name
+    // of whichever is found.
+    let path = match env::var_os("PATH") {
+        Some(x) => x,
+        None => return None
+    };
+
+    for mut dir in env::split_paths(&path) {
+        for bin in &["nodejs", "node"] {
+            dir.push(bin);
+            if dir.exists() {
+                return Some(bin.to_string());
+            }
+            dir.pop();
+        }
+    }
+    None
+}
```

Comment:
This may also be expressible as:

```rust
let path = env::var_os("PATH").unwrap_or(OsString::new());
env::split_paths(&path).flat_map(|p| {
    vec![p.join("nodejs"), p.join("node"), p.join("nodejs.exe"), p.join("node.exe")]
}).filter(|p| {
    p.exists()
}).next()
```

It's ok to not really optimize this much but we probably want to also handle windows as well?